File: GitHubPop/app/src/main/java/com/igormeira/githubpop/repository/RepositoryDataSource.java
package com.igormeira.githubpop.repository;

import androidx.annotation.NonNull;
import androidx.paging.PageKeyedDataSource;

import com.igormeira.githubpop.apirequests.GitHubClient;
import com.igormeira.githubpop.apirequests.GitHubService;
import com.igormeira.githubpop.model.Repositories;
import com.igormeira.githubpop.model.Repository;

import java.util.ArrayList;
import java.util.List;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class RepositoryDataSource extends PageKeyedDataSource<Long, Repository> {
    public static int PAGE_SIZE = 30;
    public static long INITIAL_PAGE = 1;

    @Override
    public void loadInitial(@NonNull LoadInitialParams<Long> params, @NonNull LoadInitialCallback<Long, Repository> callback) {
        GitHubService service = new GitHubClient().getGitHubService();
        Call<Repositories> call = service.getRepositories(INITIAL_PAGE);
        call.enqueue(new Callback<Repositories>() {
            @Override
            public void onResponse(Call<Repositories> call, Response<Repositories> response) {
                Repositories repositoriesBody = response.body();
                if (repositoriesBody != null && repositoriesBody.getRepositories() != null) {
                    List<Repository> repositories = (ArrayList<Repository>) repositoriesBody.getRepositories();
                    callback.onResult(repositories, null, INITIAL_PAGE + 1);
                }
            }

            @Override
            public void onFailure(Call<Repositories> call, Throwable t) {

            }
        });
    }

    @Override
    public void loadBefore(@NonNull LoadParams<Long> params, @NonNull LoadCallback<Long, Repository> callback) {
        GitHubService service = new GitHubClient().getGitHubService();
        Call<Repositories> call = service.getRepositories(params.key);
        call.enqueue(new Callback<Repositories>() {
            @Override
            public void onResponse(Call<Repositories> call, Response<Repositories> response) {
                Repositories repositoriesBody = response.body();
                if (repositoriesBody != null && repositoriesBody.getRepositories() != null) {
                    List<Repository> repositories = repositoriesBody.getRepositories();
                    long key;
                    if (params.key > 1) {
                        key = params.key - 1;
                    } else {
                        key = 0;
                    }
                    callback.onResult(repositories, key);
                }
            }

            @Override
            public void onFailure(Call<Repositories> call, Throwable t) {

            }
        });
    }

    @Override
    public void loadAfter(@NonNull LoadParams<Long> params, @NonNull LoadCallback<Long, Repository> callback) {
        GitHubService service = new GitHubClient().getGitHubService();
        Call<Repositories> call = service.getRepositories(params.key);
        call.enqueue(new Callback<Repositories>() {
            @Override public void onResponse(Call<Repositories> call, Response<Repositories> response) {
                Repositories repositoriesBody = response.body();
                if (repositoriesBody != null && repositoriesBody.getRepositories() != null) {
                    List<Repository> repositories = repositoriesBody.getRepositories();
                    callback.onResult(repositories, params.key + 1);
                }
            }
            @Override public void onFailure(Call<Repositories> call, Throwable t) {
            }
        });

    }
}

File: app/src/main/java/com/accenture/desafioandroid/data/datasource/RepositoryDataSource.kt
package com.accenture.desafioandroid.data.datasource
import android.arch.paging.PageKeyedDataSource
import android.util.Log
import com.accenture.desafioandroid.data.network.Network
import retrofit2.Call
import retrofit2.Callback
import retrofit2.HttpException
import retrofit2.Response
import com.accenture.desafioandroid.mvvm.model.Item
import com.accenture.desafioandroid.mvvm.model.RepositoryResponse


class RepositoryDataSource: PageKeyedDataSource<Int, Item>() {

    override fun loadInitial(params: LoadInitialParams<Int>, callback: LoadInitialCallback<Int, Item>) {
        val page = 1
        val inter = Network.create()
        Log.i("RepositoryDataSource","calling RepositoryDataSource")
        inter.getRepository("Java","stars", page).enqueue(object : Callback<RepositoryResponse> {
            override fun onResponse(call: Call<RepositoryResponse>, response: Response<RepositoryResponse>) {
                val searchModel = response.body()

                if (searchModel == null) {
                    onFailure(call, HttpException(response))
                    return
                }

                callback.onResult(response.body()!!.items!!, 0, searchModel.totalCount!!, null, page + 1)
            }

            override fun onFailure(call: Call<RepositoryResponse>, t: Throwable) {
            }
        })
    }

    override fun loadAfter(params: LoadParams<Int>, callback: LoadCallback<Int, Item>) {
        val page = params.key
        val inter = Network.create()
        inter.getRepository("Java","stars", page).enqueue(object : Callback<RepositoryResponse> {
            override fun onResponse(call: Call<RepositoryResponse>, response: Response<RepositoryResponse>) {
                val searchModel = response.body()

                if (searchModel ==
                    null
                ) {
                    onFailure(call, HttpException(response))
                    return
                }

                callback.onResult(
                    searchModel.items!!, // List of data items
                    // Next Page key (Used at the next request). Return `null` if this is the last page.
                    page + 1
                )
            }

            override fun onFailure(call: Call<RepositoryResponse>, t: Throwable) {
            }
        })


    }

    override fun loadBefore(params: LoadParams<Int>, callback: LoadCallback<Int, Item>) {
        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
    }

}